src/dns.rs

```diff
@@ -0,0 +1,464 @@
+use crate::topology::config::GlobalOptions;
+use futures::{Async, Future, Poll};
+use hyper::client::connect::dns::{Name, Resolve};
+use std::io;
+use std::net::ToSocketAddrs;
+use std::net::*;
+use std::vec::IntoIter;
+use trust_dns_resolver::config::*;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_resolver::AsyncResolver;
+use trust_dns_resolver::BackgroundLookupIp;
+
+/// Default port for DNS service.
+const DNS_PORT: u16 = 53;
+
+/// Default version will behave identically to std::net::ToSocketAddrs implementations.
+#[derive(Clone, Default)]
+pub struct DnsResolver(Option<AsyncResolver>);
+
+impl DnsResolver {
+    /// (Self,errors)
+    pub fn new(config: &GlobalOptions) -> Result<Self, String> {
+        if config.dns.is_empty() {
+            return Ok(DnsResolver::default());
+        }
+
+        let mut resolve_config = ResolverConfig::new();
+
+        let mut errors = Vec::new();
+
+        for s in config.dns.iter() {
+            let parsed = s
+                .parse::<SocketAddr>()
+                .or_else(|_| s.parse::<IpAddr>().map(|ip| SocketAddr::new(ip, DNS_PORT)));
+            match parsed {
+                Ok(socket_addr) => resolve_config.add_name_server(NameServerConfig {
+                    socket_addr,
+                    protocol: Protocol::Udp,
+                    tls_dns_name: None,
+                }),
+                Err(error) => errors.push(format!(
+                    "DNS server IP {:?} is invalid. Parse Error: {:?}",
+                    s, error
+                )),
+            }
+        }
+
+        if !errors.is_empty() {
+            return Err(format!(
+                "DNS resolver initialization failed with errors: {:?}",
+                errors
+            ));
+        }
+
+        let mut options = ResolverOpts::default();
+        options.attempts = 2;
+        options.validate = false;
+        options.ip_strategy = LookupIpStrategy::Ipv4AndIpv6;
+        options.use_hosts_file = false;
+        options.distrust_nx_responses = false;
```

Comment:
What are the repercussions on this?

Reply:
In situations where requested name isn't mapped, unnecessary requests to other DNS servers will be made. 

But it prevents a poorly configured or malicious DNS server to screw with DNS resolution in that way. 

Also when DNSSEC is enabled, this could be enabled.

Reply:
yeah, this mostly has implications with DNSSEC, if not using DNSSEC it's somewhat irrelevant, since your DNS might be messed with regardless and there isn't a good way to prevent or detect that. If you're operating in an environment where authenticity of the records is important, then you'd want to enable DNSSEC and/or, use one of DNS-over-TLS or DNS-over-HTTPS (DoT or DoH).

Reply:
Then do you think it's worth not to trust negative responses, so to reduce the damage of misconfigured DNS servers?

Reply:
Ok, I just reviewed the code and the background on this option to make sure I'm being accurate, and of course I'm slightly wrong, here's the code:

https://github.com/bluejekyll/trust-dns/blob/master/crates/resolver/src/name_server/name_server.rs#L147

and it relates to this issue, not DNSSEC, please ignore my early comment:

https://github.com/bluejekyll/trust-dns/issues/606

Reviewing that issue, I realize that the issue here is that where people have upstream resolvers (generally caching resolvers) that are incorrectly configured to and return NXDOMAIN instead of NoError and empty address list. Because this is a frequent enough issue, the option is defaulted to true, which will tell trust-dns to continue to the next name_server and treat an NXDOMAIN as a retriable error condition, rather than treat NXDOMAIN (correctly according to the RFC) as a non existent domain.

My recommendation is generally to go with the default, and I would suggest that here in this case.